prometheus/promdb/promdb.go

```diff
@@ -0,0 +1,132 @@
+package promdb
+
+import (
+	"database/sql"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// DBWithStats is an interface used by DatabaseCollector
+type DBWithStats interface {
+	Stats() sql.DBStats
+}
+
+// DatabaseCollector implements prometheus Collector
+type DatabaseCollector struct {
+	db     DBWithStats
+	labels prometheus.Labels
+
+	maxOpenConnections *prometheus.Desc
+	openConnections    *prometheus.Desc
+	inUse              *prometheus.Desc
+	idle               *prometheus.Desc
+	waitCount          *prometheus.Desc
+	waitDuration       *prometheus.Desc
+	maxIdleClosed      *prometheus.Desc
+	maxLifetimeClosed  *prometheus.Desc
+}
+
+// With returns a new instance of DatabaseCollector with the specified labels
+// merging with any previous labels that may have been defined
+func (c *DatabaseCollector) With(labels prometheus.Labels) *DatabaseCollector {
+	// Initialize with at least as many labels as we already have, the provided
+	// labels could overwrite current labels, so we can't be sure of the ending
+	// size.
+	newLabels := make(prometheus.Labels, len(c.labels))
+	for k, v := range c.labels {
+		newLabels[k] = v
+	}
+	for k, v := range labels {
+		newLabels[k] = v
+	}
+	return NewDatabaseCollectorWithLabels(c.db, newLabels)
+}
+
+// WithLabel is a shortcut to With for providing a single label and value
+func (c *DatabaseCollector) WithLabel(name, value string) *DatabaseCollector {
+	labels := prometheus.Labels{name: value}
+	return c.With(labels)
+}
+
+// Describe implements prometheus Collector
+func (c *DatabaseCollector) Describe(ch chan<- *prometheus.Desc) {
```

Comment:
I assume this pattern was figured out from a prometheus doc? Would be good to link to it to know where this non-obvious-send-pattern comes from (in the commit message maybe, but could be in line too)

Reply:
I've added some additional details (that it's implementing an interface) but I think this should be relatively obvious that you can find it in their docs.